plugin/gcp/pubsubTopic.go

```diff
@@ -0,0 +1,160 @@
+package gcp
+
+import (
+	"container/list"
+	"context"
+	"fmt"
+	"io"
+	"runtime"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/puppetlabs/wash/plugin"
+	"google.golang.org/api/iterator"
+)
+
+type pubsubTopic struct {
+	plugin.EntryBase
+	client *pubsub.Client
+	topic  *pubsub.Topic
+}
+
+type pubsubTopicMeta struct {
+	pubsub.PublishSettings
+	pubsub.TopicConfig
+	Subscriptions []string
+}
+
+func newPubsubTopic(client *pubsub.Client, topic *pubsub.Topic) *pubsubTopic {
+	top := &pubsubTopic{
+		EntryBase: plugin.NewEntry(topic.ID()),
+		client:    client,
+		topic:     topic,
+	}
+	top.
+		Attributes().
+		SetMeta(pubsubTopicMeta{PublishSettings: topic.PublishSettings})
+
+	// This may be somewhat hacky, but it ensures the goroutines for publishing get cleaned up eventually.
+	// Writeable can optionally also implement io.Closer to do cleanup when we're done writing,
+	// but we may re-use this client so we shouldn't stop it in Close.
+	runtime.SetFinalizer(top, func(t *pubsubTopic) { t.topic.Stop() })
+	return top
+}
+
+func (t *pubsubTopic) Metadata(ctx context.Context) (plugin.JSONObject, error) {
+	cfg, err := t.topic.Config(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	subs := make([]string, 0)
+	it := t.topic.Subscriptions(ctx)
+	for {
+		s, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		subs = append(subs, s.ID())
+	}
+
+	return plugin.ToJSONObject(pubsubTopicMeta{
+		PublishSettings: t.topic.PublishSettings,
+		TopicConfig:     cfg,
+		Subscriptions:   subs,
+	}), nil
+}
+
+func (t *pubsubTopic) Delete(ctx context.Context) (bool, error) {
+	return true, t.topic.Delete(ctx)
+}
+
+// A ReadCloser that subscribes to a topic and buffers all messages that appear there.
+type pubsubTopicWatcher struct {
+	*pubsub.Subscription
+	sync.Mutex
+	*list.List
+	context.Context
+	context.CancelFunc
+	err error
+}
+
+func newPubsubTopicWatcher(sub *pubsub.Subscription) *pubsubTopicWatcher {
```

Comment:
Nit: This could be a little cleaner as `t.newWatcher() (*pubsubTopicWatcher, error)` because then you can put creation of the temp. subscription in there. It's not a huge deal though.